### PR TITLE
JSON output fetching fixes

### LINKED ISSUE #1453: [terraform/output] invalid character 'm' after array element
**Describe the bug**
When calling the `OutputRequired` function in the `terraform/output` module, I am getting an Error: `Received unexpected error: invalid character 'm' after array element`.

This error looks to have been reported in this [Issue](https://github.com/gruntwork-io/terratest/issues/1442), but continues to persist.  

**To Reproduce**
Run the following [internal test](https://github.com/gruntwork-io/terraform-aws-ci/blob/main/test/destroy_test.go#L47)
[Offending line ](https://github.com/gruntwork-io/terraform-aws-ci/blob/main/test/destroy_test.go#L240)

Result:
```sh
--- FAIL: TestIntegrationDestroy (537.03s)
    output.go:36: 
        	Error Trace:	/Users/me/go/pkg/mod/github.com/gruntwork-io/terratest@v0.47.2/modules/terraform/output.go:36
        	            				/Users/me/src/github.com/gruntwork/terraform-aws-ci/test/destroy_test.go:240
        	            				/Users/me/go/pkg/mod/github.com/gruntwork-io/terratest@v0.47.2/modules/test-structure/test_structure.go:30
        	            				/Users/me/src/github.com/gruntwork/terraform-aws-ci/test/destroy_test.go:207
        	Error:      	Received unexpected error:
        	            	invalid character 'm' after array element
        	Test:       	TestIntegrationDestroy
FAIL
```

**Expected behavior**

**Nice to have**
- [ ] Terminal output
- [ ] Screenshots

**Versions**
- Terratest version:
- Environment details (Ubuntu 20.04, Windows 10, etc.):

**Additional context**
- `terratest` versions `v0.47.1` and `v0.47.2`
- MacOS `14.2.1 (23C71)`

## When debugging locally
```go

// line 299 of output.go
OutputStructE(out)
// input value of out as input to function cleanOutput(out)
out = "\x1b[90m09:06:11.281 \x1b[0m\x1b[32mINFO   \x1b[0mDownloading Terraform configurations from ../../../../fixtures/null_resource_module into ./.terragrunt-cache/db4UlyBp-l9tLecIJQhY4ihzvZo/WTtZRBNSmdc6CoLXtNhzYg9-r0Q\n\x1b[90m09:06:11.311 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mInitializing the backend...\n\x1b[90m09:06:11.318 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mSuccessfully configured the backend \"local\"! Terraform will automatically\n\x1b[90m09:06:11.318 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0muse this backend unless the backend configuration changes.\n\x1b[90m09:06:11.318 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mInitializing provider plugins...\n\x1b[90m09:06:11.318 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0m- Finding hashicorp/aws versions matching \"< 4.0.0\"...\n\x1b[90m09:06:11.641 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0m- Finding latest version of hashicorp/null...\n\x1b[90m09:06:12.074 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0m- Installing hashicorp/null v3.2.3...\n\x1b[90m09:06:12.429 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0m- Installed hashicorp/null v3.2.3 (signed by HashiCorp)\n\x1b[90m09:06:12.610 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0m- Installing hashicorp/aws v3.76.1...\n\x1b[90m09:06:14.625 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0m- Installed hashicorp/aws v3.76.1 (signed by HashiCorp)\n\x1b[90m09:06:14.625 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mTerraform has created a lock file .terraform.lock.hcl to record the provider\n\x1b[90m09:06:14.625 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mselections it made above. Include this file in your version control repository\n\x1b[90m09:06:14.625 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mso that Terraform can guarantee to make the same selections by default when\n\x1b[90m09:06:14.625 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0myou run \"terraform init\" in the future.\n\x1b[90m09:06:14.625 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mTerraform has been successfully initialized!\n\x1b[90m09:06:14.626 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mYou may now begin working with Terraform. Try running \"terraform plan\" to see\n\x1b[90m09:06:14.626 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0many changes that are required for your infrastructure. All Terraform commands\n\x1b[90m09:06:14.626 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mshould now work.\n\x1b[90m09:06:14.626 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mIf you ever set or change modules or backend configuration for Terraform,\n\x1b[90m09:06:14.626 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mrerun this command to reinitialize your working directory. If you forget, other\n\x1b[90m09:06:14.626 \x1b[0m\x1b[37mSTDOUT \x1b[0m\x1b[36mterraform: \x1b[0mcommands will detect it and remind you to do so if necessary.\n\"3919159218922798988\"\n\x1b[90m09:06:14.815 \x1b[0m\x1b[32mINFO   \x1b[0mRetrieved output from terraform"

//return value of out after running from cleanOutput(out)
out = "[90m09:06:11.311 [0m[37mSTDOUT [0m[36mterraform: [0mInitializing the backend...[90m09:06:11.318 [0m[37mSTDOUT [0m[36mterraform: [0mSuccessfully configured the backend \"local\"! Terraform will automatically[90m09:06:11.318 [0m[37mSTDOUT [0m[36mterraform: [0muse this backend unless the backend configuration changes.[90m09:06:11.318 [0m[37mSTDOUT [0m[36mterraform: [0mInitializing provider plugins...[90m09:06:11.318 [0m[37mSTDOUT [0m[36mterraform: [0m- Finding hashicorp/aws versions matching \"< 4.0.0\"...[90m09:06:11.641 [0m[37mSTDOUT [0m[36mterraform: [0m- Finding latest version of hashicorp/null...[90m09:06:12.074 [0m[37mSTDOUT [0m[36mterraform: [0m- Installing hashicorp/null v3.2.3...[90m09:06:12.429 [0m[37mSTDOUT [0m[36mterraform: [0m- Installed hashicorp/null v3.2.3 (signed by HashiCorp)[90m09:06:12.610 [0m[37mSTDOUT [0m[36mterraform: [0m- Installing hashicorp/aws v3.76.1...[90m09:06:14.625 [0m[37mSTDOUT [0m[36mterraform: [0m- Installed hashicorp/aws v3.76.1 (signed by HashiCorp)[90m09:06:14.625 [0m[37mSTDOUT [0m[36mterraform: [0mTerraform has created a lock file .terraform.lock.hcl to record the provider[90m09:06:14.625 [0m[37mSTDOUT [0m[36mterraform: [0mselections it made above. Include this file in your version control repository[90m09:06:14.625 [0m[37mSTDOUT [0m[36mterraform: [0mso that Terraform can guarantee to make the same selections by default when[90m09:06:14.625 [0m[37mSTDOUT [0m[36mterraform: [0myou run \"terraform init\" in the future.[90m09:06:14.625 [0m[37mSTDOUT [0m[36mterraform: [0mTerraform has been successfully initialized![90m09:06:14.626 [0m[37mSTDOUT [0m[36mterraform: [0mYou may now begin working with Terraform. Try running \"terraform plan\" to see[90m09:06:14.626 [0m[37mSTDOUT [0m[36mterraform: [0many changes that are required for your infrastructure. All Terraform commands[90m09:06:14.626 [0m[37mSTDOUT [0m[36mterraform: [0mshould now work.[90m09:06:14.626 [0m[37mSTDOUT [0m[36mterraform: [0mIf you ever set or change modules or backend configuration for Terraform,[90m09:06:14.626 [0m[37mSTDOUT [0m[36mterraform: [0mrerun this command to reinitialize your working directory. If you forget, other[90m09:06:14.626 [0m[37mSTDOUT [0m[36mterraform: [0mcommands will detect it and remind you to do so if necessary.\"3919159218922798988\""

// fails on error thrown from this line
return json.Unmarshal([]byte(out), &v)

```

### LINKED ISSUE #1461: terraform/output: unicode characters missing since 0.47.2
**Describe the bug**
I have a test that calls `terraform.Output` and that output contains a unicode character. After updating terratest from 0.47.1 to 0.47.2, those unicode characters are missing.

**To Reproduce**
Steps to reproduce the behavior, code snippets and examples which can be used to reproduce the issue.

```terraform
output "id" {
  value = "söme chäräcter"
}
```

```go
assert.Equal(t, "söme chäräcter", terraform.Output(t, options, "id"))
```

**Expected behavior**
The given output should be given as-is without removing unicode characters.

**Nice to have**
- See job logs at https://github.com/metio/terraform-provider-migadu/actions/runs/11130950971/job/30931610454?pr=205

**Versions**
- Terratest version: 0.47.2
- Environment details (Ubuntu 20.04, Windows 10, etc.): Ubuntu

**Additional context**

I think this is caused by https://github.com/gruntwork-io/terratest/pull/1445 which keeps printable ASCII chars only

### PR DESCRIPTION
**Pull request by denis256**
## Description

Included changes:
* updated Terragrunt version to latest to detect output parsing errors
* updated Terragrunt log level flags to redirect output
* added cleanup and formatting of json output
* added tests to track output fetching with Terragrunt
* added test to track unicode characters handling

Fixes #1453.
Fixes #1461.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

Updated output parsing from Terragrunt

### CODE REVIEW

**Comment by denis256** [comment]
terraform-aws-ci tests are passing too

image

File: modules/terraform/cmd.go [hunk 1]
```diff
@@ -49,8 +52,16 @@ func GetCommonOptions(options *Options, args ...string) (*Options, []string) {
...
-	if options.TerraformBinary == "terragrunt" {
+	if options.TerraformBinary == TerragruntDefaultPath {
 		args = append(args, "--terragrunt-non-interactive")
+		// for newer Terragrunt version, setting simplified log formatting
+		if options.EnvVars == nil {
+			options.EnvVars = map[string]string{}
+		}
+		_, tgLogSet := options.EnvVars["TERRAGRUNT_DISABLE_LOG_FORMATTING"]
+		if !tgLogSet {
+			options.EnvVars["TERRAGRUNT_DISABLE_LOG_FORMATTING"] = "true"
```
Comment (levkohimins) [inline modules/terraform/cmd.go:63]: This env var is deprecated, please use `TERRAGRUNT_LOG_FORMAT=key-value`

*But only if you use the latest version that supports log format.

**Review by levkohimins — DISMISSED** [review]
Let me know if my suggestions are not relevant, I will approve it.

File: modules/terraform/output.go [hunk 2]
```diff
@@ -5,12 +5,21 @@ import (
...
 
 	"github.com/gruntwork-io/terratest/modules/testing"
 	"github.com/stretchr/testify/require"
 )
 
+const skipJsonLogLine = " msg="
+
+var (
+	ansiLineRegex = regexp.MustCompile(`(?m)^\x1b\[[0-9;]*m.*`)
+	tgLogLevel    = regexp.MustCompile(`.*time=\S+ level=\S+ prefix=\S+ binary=\S+ msg=.*`)
```
[the remark this answers is not in the document]
Comment (denis256) [reply modules/terraform/output.go:20]: Output still has lines like:
```
time=2024-12-05UTC17:32:10Z level=stdout tf-path=terraform msg=Initializing the backend...
```

which require cleaning

File: modules/terraform/cmd.go [hunk 3]
```diff
@@ -49,8 +52,20 @@ func GetCommonOptions(options *Options, args ...string) (*Options, []string) {
...
-	if options.TerraformBinary == "terragrunt" {
+	if options.TerraformBinary == TerragruntDefaultPath {
 		args = append(args, "--terragrunt-non-interactive")
+		// for newer Terragrunt version, setting simplified log formatting
+		if options.EnvVars == nil {
+			options.EnvVars = map[string]string{}
+		}
+		_, tgLogSet := options.EnvVars["TERRAGRUNT_LOG_FORMAT"]
+		if !tgLogSet {
+			options.EnvVars["TERRAGRUNT_LOG_FORMAT"] = "key-value"
```
Comment (james03160927) [inline modules/terraform/cmd.go:63]: Can you explain what this does? It could be great to add some comments explanining the default behaviour here and below.

File: modules/terraform/output.go [hunk 4]
```diff
@@ -5,12 +5,21 @@ import (
...
+	"strings"
 
 	"github.com/gruntwork-io/terratest/modules/testing"
 	"github.com/stretchr/testify/require"
 )
 
+const skipJsonLogLine = " msg="
+
+var (
+	ansiLineRegex = regexp.MustCompile(`(?m)^\x1b\[[0-9;]*m.*`)
```
Comment (james03160927) [inline modules/terraform/output.go:19]: nit: can we add comments on what these regex mean in plain English?

**Review by james03160927 — APPROVED** [review]
A few nit comments but overall LGTM.

File: modules/terraform/output.go (hunk 4 above)
Comment (denis256) [reply modules/terraform/output.go:20]: Updated

File: modules/terraform/cmd.go (hunk 3 above)
Comment (denis256) [reply modules/terraform/cmd.go:65]: Added description

**Review by james03160927 — APPROVED** [review]
LGTM
